doc: moved the doc to include/drivers, language review (#1608)
NordicPlayground/fw-nrfconnect-nrf: include/drivers/at_cmd.rst

﻿.. _at_cmd_readme:

AT command interface
####################

The socket interface has a global limitation of 8 socket instances that can be

> open at the same time.

Author: Removed.

So if an application needs multiple socket protocols like TCP,
UDP, TLS, AT, and so on, the available 8 socket instances become a limiting factor.
The AT command interface driver is designed to let multiple clients, i.e.,

> threads, software modules/components to share a single AT socket instance.

Reviewer: The important part here is actually just "threads". So maybe it is enough to say ".... multiple threads share ....", and drop the content between the parenthesis. Because that is what must be synchronized.

> threads, software modules/components to share a single AT socket instance.

Author: Updated.

The AT command driver always parses the return code from the modem.

> Non-notification data, if any, such as OK, ERROR, and +CMS/+CME are removed from the string returned to the user.

Author: Modified.

> The return codes are returned as error codes in the return code of the write function

Reviewer: Yes

> The return codes are returned as error codes in the return code of the write function

Author: Added links to both the functions.

and also through the state parameter that can be supplied.
The state parameter must be used to differentiate between +CMS and +CME errors
as the error codes are overlapping.
Any subsequent writes from other threads are queued until all the data (return
code + any payload) from the previous write is returned to the caller.
This is to make sure that the correct thread gets the correct data and the return

> code as there is no way to separate two sessions apart.

Author: Modified.

The kernel thread priority scheme determines the thread to get write access if multiple threads are queued.

There are two schemes by which data returned immediately from the modem (for instance, the modem response for an AT+CNUM command) is delivered to the user.

> The user can submit either a reference to a string buffer or a reference to a handler function in the write function.

Reviewer 2: Doesn't the user need to call different functions, depending on how they want the data returned? This sentence makes it sound like the distinction is made based on what the user submits to one function.

Author: Modified as:
 The user can call the write function by submitting either of the following in the write function:...

In the case of a user string buffer, the driver removes the return code and then delivers the rest of the string
in the buffer if it is large enough.
Data is returned to the user if the return code is OK.

> In the case of a handler function, the return code is removed and the rest of

Reviewer 2: This and line 27 are two options. They could be changed to bullet points, or to one paragraph or two paragraphs.

Author: Changed to bullets, followed by two paragraphs, each paragraph explaining an option.

the string is delivered to the handler function through a char pointer
parameter.
Allocation and deallocation of the buffer is handled by the driver,
and the content should not be considered valid outside of the handler.
Both schemes are limited to the maximum reception size defined by

> CONFIG_AT_CMD_RESPONSE_MAX_LEN.

Reviewer 2: ```suggestion
:option:`CONFIG_AT_CMD_RESPONSE_MAX_LEN`.
```

Author: Updated

> Notifications are always handled by a callback function and it is separate from

Reviewer 2: ```suggestion
Notifications are always handled by a callback function that is separate from
```

Author: changed.
